### PR TITLE
did more work on genetic algorithm

### PR DESCRIPTION
**Pull request by amccarthy9904**

### CODE REVIEW

File: NeuralNetworkTrainer/src/neuralNetworkTrainer/Driver.java [hunk 1]
```diff
@@ -47,6 +48,10 @@
 	static double mutationRate; // used by GA and ES
 	static double beta;         // used by DE
 
+	/**
+	 * Random number generator
+	 */
+	public static Random randNum = new Random();
```
Comment (EliasAthey) [inline NeuralNetworkTrainer/src/neuralNetworkTrainer/Driver.java]: What's the benefit of having this in Driver? Could we just use Random when it's needed

File: NeuralNetworkTrainer/src/neuralNetworkTrainer/EvolutionStrategy.java [hunk 2]
```diff
@@ -1,15 +1,25 @@
...
+		
+		ArrayList<Network> population = null;
+				
+		//create populationSize number of individuals and add them to population
+		for(int popIter = 0;  popIter < Driver.populationSize; popIter++) {
+			
+			//adding a global config in driver might be worth doing
+			//or passing config through train() to all these methods
+			Network individual = new Network(Driver.configuration);
+			population.add(individual);
```
Comment (EliasAthey) [inline NeuralNetworkTrainer/src/neuralNetworkTrainer/EvolutionStrategy.java]: I think our population should be `ArrayList<ArrayList<Double>>` objects representing the weights

File: NeuralNetworkTrainer/src/neuralNetworkTrainer/GeneticAlgorithm.java [hunk 3]
```diff
@@ -1,33 +1,161 @@
...
+		}
+		
+		return deserializedPopulation;
 	}
 	
-	public ArrayList<Network> crossoverOffspring(ArrayList<Network> offspring){
-		//TODO
-		return null;
+	public ArrayList<ArrayList<ArrayList<Double>>> serializePopulation( ArrayList<Network> population){
+		ArrayList<ArrayList<ArrayList<Double>>> serializedPopulation = new ArrayList<ArrayList<ArrayList<Double>>>();
```
Comment (EliasAthey) [inline NeuralNetworkTrainer/src/neuralNetworkTrainer/GeneticAlgorithm.java]: You can change this to `ArrayList<ArrayList<ArrayList<Double>>> serializedPopulation = new ArrayList<>();`

File: NeuralNetworkTrainer/src/neuralNetworkTrainer/GeneticAlgorithm.java [hunk 4]
```diff
@@ -1,33 +1,161 @@
...
+			for(int innerIter = 0; innerIter < parent1.get(0).size(); innerIter++) {
+				
+				if (randomizer.get(outerIter).get(innerIter)) {
+					offspring1.get(outerIter).add(parent1.get(outerIter).get(innerIter));
+					offspring2.get(outerIter).add(parent2.get(outerIter).get(innerIter));
+				}
+				
+				else {//flip which offspring get which gene
+					offspring2.get(outerIter).add(parent1.get(outerIter).get(innerIter));
+					offspring2.get(outerIter).add(parent2.get(outerIter).get(innerIter));
```
Comment (EliasAthey) [inline NeuralNetworkTrainer/src/neuralNetworkTrainer/GeneticAlgorithm.java]: Should assign offspring1 in second line

File: NeuralNetworkTrainer/src/neuralNetworkTrainer/GeneticAlgorithm.java [hunk 5]
```diff
@@ -1,33 +1,161 @@
...
+
+	public ArrayList<ArrayList<ArrayList<Double>>> mutateOffspring(ArrayList<ArrayList<ArrayList<Double>>> offspring){
+		for (ArrayList<ArrayList<Double>> individual : offspring) {
+
+			for (ArrayList<Double> chromosome : individual) {
+				for(Double gene : chromosome) {
+					if (Math.random() <= Driver.mutationRate) { //gives a mustationRate % chance of a mutation happening on any given gene, lower for higher fitness individuals maybe????
+						//change gene with a random number from a gaussian distribution 
+						//centered at 0 with a standard deviation of 1
+						gene += Driver.randNum.nextGaussian();
```
Comment (EliasAthey) [inline NeuralNetworkTrainer/src/neuralNetworkTrainer/GeneticAlgorithm.java]: I think GA uses a Normal distribution centered at 0 or the gene's original value. The spread of the distribution is the standard deviation of the gene. So we need to 1) track the std. dev. of each gene, and 2) multiply `Driver.randNum.nextGaussian()` by the std. dev.

Comment (EliasAthey) [reply NeuralNetworkTrainer/src/neuralNetworkTrainer/GeneticAlgorithm.java]: To clarify, nextGaussian() returns a normal distribution centered at 0 with a std. dev. of 1, so we need to scale it.

File: NeuralNetworkTrainer/src/neuralNetworkTrainer/GeneticAlgorithm.java [hunk 6]
```diff
@@ -36,10 +164,41 @@ public Boolean hasConverged (ArrayList<Network> currentPopulation, ArrayList<Net
 		return null;
 	}
 	
+	//maybe we can pass this a boolean to differentiate between classification and function approx
+	//we would only need to check the boolean for eval fitness, maybe hasConverged
```
Comment (EliasAthey) [inline NeuralNetworkTrainer/src/neuralNetworkTrainer/GeneticAlgorithm.java:229]: Driver.isClassificationNetwork is that boolean

File: NeuralNetworkTrainer/src/neuralNetworkTrainer/GeneticAlgorithm.java [hunk 7]
```diff
@@ -36,10 +164,41 @@ public Boolean hasConverged (ArrayList<Network> currentPopulation, ArrayList<Net
...
-		return null;
+		// TODO 
+		ArrayList<Network> prevPopulation = null;
+		ArrayList<Network> offspring =null;
+		ArrayList<ArrayList<ArrayList<Double>>> serializedPopulation = new ArrayList<ArrayList<ArrayList<Double>>>();
+		ArrayList<ArrayList<ArrayList<Double>>> serializedOffspring = new ArrayList<ArrayList<ArrayList<Double>>>();
+
+		ArrayList<Network> population = generatePopulation();
+		population = evalClasificationFitness(population);
+		while(hasConverged(population, prevPopulation)){
```
Comment (EliasAthey) [inline NeuralNetworkTrainer/src/neuralNetworkTrainer/GeneticAlgorithm.java]: Add a !, we want to loop while we DONT converge

File: NeuralNetworkTrainer/src/neuralNetworkTrainer/GeneticAlgorithm.java [hunk 8]
```diff
@@ -36,10 +164,41 @@ public Boolean hasConverged (ArrayList<Network> currentPopulation, ArrayList<Net
...
+			serializedPopulation = serializePopulation(population);
+			prevPopulation = population;
+			serializedOffspring = newGeneration(serializedPopulation);
+			offspring = deserializePopulation(serializedOffspring);
+			offspring = evalClasificationFitness(offspring);
+		//	replacePop(offspring, offspring.fitness)
+		//}
+		//return population;
+		return new Network(null);
+		}
```
Comment (EliasAthey) [inline NeuralNetworkTrainer/src/neuralNetworkTrainer/GeneticAlgorithm.java]: train() needs to return a network

File: NeuralNetworkTrainer/src/neuralNetworkTrainer/Driver.java (hunk 1 above)
Comment (amccarthy9904) [reply NeuralNetworkTrainer/src/neuralNetworkTrainer/Driver.java]: I dont like declaring things that are not super important to the logic inside methods. We could make it an instance variable.

File: NeuralNetworkTrainer/src/neuralNetworkTrainer/GeneticAlgorithm.java (hunk 6 above)
Comment (amccarthy9904) [reply NeuralNetworkTrainer/src/neuralNetworkTrainer/GeneticAlgorithm.java:229]: oh sweet

Comment (amccarthy9904) [reply NeuralNetworkTrainer/src/neuralNetworkTrainer/GeneticAlgorithm.java:229]: done

File: NeuralNetworkTrainer/src/neuralNetworkTrainer/EvolutionStrategy.java (hunk 2 above)
Comment (amccarthy9904) [reply NeuralNetworkTrainer/src/neuralNetworkTrainer/EvolutionStrategy.java]: yeah i havent been keeping up with changing ES and DE. i can change that stuff though

File: NeuralNetworkTrainer/src/neuralNetworkTrainer/GeneticAlgorithm.java (hunk 4 above)
Comment (amccarthy9904) [reply NeuralNetworkTrainer/src/neuralNetworkTrainer/GeneticAlgorithm.java]: good catch

File: NeuralNetworkTrainer/src/neuralNetworkTrainer/GeneticAlgorithm.java (hunk 5 above)
Comment (amccarthy9904) [reply NeuralNetworkTrainer/src/neuralNetworkTrainer/GeneticAlgorithm.java]: ok i implemented this. its kinda messy though

File: NeuralNetworkTrainer/src/neuralNetworkTrainer/GeneticAlgorithm.java (hunk 7 above)
Comment (amccarthy9904) [reply NeuralNetworkTrainer/src/neuralNetworkTrainer/GeneticAlgorithm.java]: done

File: NeuralNetworkTrainer/src/neuralNetworkTrainer/GeneticAlgorithm.java (hunk 8 above)
Comment (amccarthy9904) [reply NeuralNetworkTrainer/src/neuralNetworkTrainer/GeneticAlgorithm.java]: done

File: NeuralNetworkTrainer/src/neuralNetworkTrainer/GeneticAlgorithm.java (hunk 3 above)
Comment (amccarthy9904) [reply NeuralNetworkTrainer/src/neuralNetworkTrainer/GeneticAlgorithm.java]: done

File: NeuralNetworkTrainer/src/neuralNetworkTrainer/Driver.java (hunk 1 above)
Comment (amccarthy9904) [reply NeuralNetworkTrainer/src/neuralNetworkTrainer/Driver.java]: made it an instance variable

File: NeuralNetworkTrainer/src/neuralNetworkTrainer/EvolutionStrategy.java [hunk 9]
```diff
@@ -1,43 +1,109 @@
...
+	private ArrayList<Network> evalFunApproxFitness(ArrayList<Network> population) {
+		// TODO Auto-generated method stub
 		return null;
 	}
-	
-	public Boolean hasConverged (ArrayList<Network> currentPopulation, ArrayList<Network> prevPopulation ){
-		//TODO
+
+	private ArrayList<Network> evalClasificationFitness(ArrayList<Network> population) {
+		// TODO Auto-generated method stub
```
Comment (EliasAthey) [inline NeuralNetworkTrainer/src/neuralNetworkTrainer/EvolutionStrategy.java:98]: I think the evaluation for classification and function approx will be the same, no need for separate methods

Comment (amccarthy9904) [reply NeuralNetworkTrainer/src/neuralNetworkTrainer/EvolutionStrategy.java:98]: done

**Pull request merged**